### PR TITLE
Support the 'workspace/configuration' message from Pylance.

### PR DESCRIPTION
**Pull request by rchiodo**
Change to responding to 'workspace/configuration' so that we can have a different interpreter per project.

### CODE REVIEW

File: Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs [hunk 1]
```diff
@@ -173,9 +173,78 @@ internal sealed class PythonLanguageClient : ILanguageClient, ILanguageClientCus
...
+                            typeCheckingMode = typeCheckingMode,
+                            useLibraryCodeForTypes = true,
+                            completeFunctionParens = _advancedEditorOptions.CompleteFunctionParens,
+                            autoImportCompletions = _advancedEditorOptions.AutoImportCompletions
+                        }
+                    };
+                }
+
+                // Copy settings in order
+                var resultArray = new object[result.Length + 1];
```
Comment (rchiodo) [inline Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs:238]: I'm going to change this to use a list and then convert the list to an array at the end.

**Comment by jakebailey** [comment]
IIRC there's an initialization option that needs to be set to declare that the client supports the call; is that in this PR? I did a look but didn't find it, so wondering if it's working regardless.

File: Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs [hunk 2]
```diff
@@ -472,6 +485,7 @@ public Task<object> InvokeCommandAsync(LSP.ExecuteCommandParams request, Cancell
                                 }
                                 capabilities["workspace"]["workspaceFolders"] = true;
                                 capabilities["workspace"]["didChangeWatchedFiles"]["dynamicRegistration"] = true;
+                                capabilities["workspace"]["configuration"] = true;
```
Comment (rchiodo) [inline Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs:485]: Right here @jakebailey

Comment (rchiodo) [reply Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs:485]: That's where the setting is turned on.

Comment (jakebailey) [reply Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs:485]: Thanks, clearly I had the wrong search terms.

File: Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs [hunk 3]
```diff
@@ -173,9 +173,75 @@ internal sealed class PythonLanguageClient : ILanguageClient, ILanguageClientCus
...
+                    var typeCheckingMode = UserSettings.GetStringSetting(
+                        PythonConstants.TypeCheckingModeSetting, null, Site, PythonWorkspaceContextProvider.Workspace, out _)
+                        ?? _analysisOptions.TypeCheckingMode;
+
+                    var ver3 = new Version(3, 0);
+                    if (context.InterpreterConfiguration.Version < ver3) {
+                        MessageBox.ShowWarningMessage(Site, Strings.WarningPython2NotSupported);
+                    }
+
+                    settings = new LanguageServerSettings.PythonSettings {
```
Comment (jakebailey) [inline Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs:219]: Do you know if PYTHONPATH is supported? There's another config that could be added there if so.

**Review by int19h — APPROVED** [review]

File: Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs [hunk 4]
```diff
@@ -173,9 +173,75 @@ internal sealed class PythonLanguageClient : ILanguageClient, ILanguageClientCus
...
+                        pythonPath = context.InterpreterConfiguration.InterpreterPath,
```
Comment (rchiodo) [inline Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs:220]: This one right here? Or did you mean something else?

Comment (rchiodo) [reply Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs:220]: Did you mean this one?

```_envPYTHONPATH```

Comment (rchiodo) [reply Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs:220]: Yes I think I can add that one too. I'll do _envPYTHONPATH

Comment (rchiodo) [reply Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs:220]: Sorry it doesn't have that information. It just has the name of the 'PYTHONPATH' environment variable but not the actual values.

Comment (rchiodo) [reply Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs:220]: Is it necessary for pylance to work correctly? I think I might be able to run the environment to extract it, but that's a lot more code.

Comment (jakebailey) [reply Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs:220]: No, it's not necessary to run. If VS doesn't provide a way to configure this, then there's no reason to try and fetch it. VSC does, so I piped that down.

File: Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs [hunk 5]
```diff
@@ -298,17 +361,15 @@ public Task<object> InvokeCommandAsync(LSP.ExecuteCommandParams request, Cancell
             }
         }
 
-        private void OnSettingsChanged(object sender, EventArgs e) => SendDidChangeConfigurations().DoNotWait();
+        private void OnSettingsChanged(object sender, EventArgs e) => InvokeDidChangeConfigurationAsync(new LSP.DidChangeConfigurationParams() {
+            Settings = null
```
Comment (rchiodo) [inline Python/Product/PythonTools/PythonTools/LanguageServerClient/PythonLanguageClient.cs:365]: I think I should put a comment here indicating why this is null.

**Pull request merged**